thorlcr/activities/nsplitter/thnsplitterslave.cpp

```diff
@@ -194,45 +195,55 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
-                    smartBuf.setown(createSharedSmartDiskBuffer(this, tempname.str(), numOutputs, queryRowInterfaces(input), &container.queryJob().queryIDiskUsage()));
-                    ActPrintLog("Using temp spill file: %s", tempname.str());
+                    CSplitterOutput *output = (CSplitterOutput *)outputs.item(o);
+                    if (output->isStopped())
+                        --activeOutputCount;
                 }
+                assertex(activeOutputCount); // must be >=1, as this output (outIdx) has invoked prepareInput
+                if (1 == activeOutputCount)
+                    return false;
+                if (smartBuf)
+                    smartBuf->reset();
                 else
                 {
-                    ActPrintLog("Spill is 'balanced'");
-                    smartBuf.setown(createSharedSmartMemBuffer(this, numOutputs, queryRowInterfaces(input), NSPLITTER_SPILL_BUFFER_SIZE));
-                }
-                // mark any outputs already stopped
-                ForEachItemIn(o, outputs)
-                {
-                    CSplitterOutput *output = (CSplitterOutput *)outputs.item(o);
-                    if (output->isStopped() || !connectedOutputSet->test(o))
-                        smartBuf->queryOutput(o)->stop();
+                    if (spill)
+                    {
+                        StringBuffer tempname;
+                        GetTempName(tempname, "nsplit", true); // use alt temp dir
+                        smartBuf.setown(createSharedSmartDiskBuffer(this, tempname.str(), numOutputs, queryRowInterfaces(input), &container.queryJob().queryIDiskUsage()));
+                        ActPrintLog("Using temp spill file: %s", tempname.str());
+                    }
+                    else
+                    {
+                        ActPrintLog("Spill is 'balanced'");
+                        smartBuf.setown(createSharedSmartMemBuffer(this, numOutputs, queryRowInterfaces(input), NSPLITTER_SPILL_BUFFER_SIZE));
+                    }
+                    // mark any outputs already stopped
+                    ForEachItemIn(o, outputs)
+                    {
+                        CSplitterOutput *output = (CSplitterOutput *)outputs.item(o);
+                        if (output->isStopped() || !connectedOutputSet->test(o))
+                            smartBuf->queryOutput(o)->stop();
+                    }
                 }
+                if (!spill)
+                    writer.start(); // writer keeps writing ahead as much as possible, the readahead impl. will block when has too much
+            }
+            catch (IException *e)
+            {
+                writeAheadException.set(e);
+                throw;
             }
-            if (!spill)
-                writer.start(); // writer keeps writing ahead as much as possible, the readahead impl. will block when has too much
         }
+        if (writeAheadException)
```

Comment:
What happens if writeAheadException is set by reading rows ahead - isn't this failing too early?  It would be a problem if the exception was being caught on one branch.

Similarly I'm not sure the nextrow logic is correct - it may throw the exception too early.  E.g. with code something like:

```
ds := LIMIT(something, 2);
output(choose(ds,1));
output(catch(ds,...));

```

Reply:
>What happens if writeAheadException is set by reading rows ahead 

well, yes I think there's a problem if the async thread (CWriter) hits an exception whilst reading rows (or preparing input)
Not a new problem though..
It should probably tuck the exception behind it's ear and only throw it when the smart look ahead buffer is exhausted by a reader.

>Similarly I'm not sure the nextrow logic is correct - it may throw the exception too early.

So this change would cause the exception to be return on both branches of the splitter, even if e.g. the choosen(ds,1) wasn't the cause.. 
I guess similarly, it should only throw it on a branch if it 'reaches' the exception previous caught/tucked away.